### PR TITLE
[TransitioningContentControl] Manage his LogicalChildren

### LINKED ISSUE #12158: UserControl loses its visual children when reused by TransitioningContentControl
**Describe the bug**
`UserControl` loses its visual children when reused by `TransitioningContentControl`. The issue doesn't happen anymore if a `ControlControl` is used instead but then I lose the transition. I attached a minimal reproduction below.

**To Reproduce**

1. Run the reproduction project attached below.
image
2. Click on 'Switch view', observe that the text changed to "Hello from user control!".
image
3. Click on 'Switch view', it should be back to "Hello world! 1" again.
image
4. Click on 'Switch view', we expect "Hello from user control!" again but it's *gone*.
image

**Expected behavior**
It should show "Hello from user control!" again.

**Desktop (please complete the following information):**

- OS: Windows (tho I assume the issue is platform-agnostic)
- Version 11.0.0

**Additional context**

[TestAvalonia.zip](https://github.com/AvaloniaUI/Avalonia/files/12027404/TestAvalonia.zip)

The `TestUserControl` contains code that prints out when it is attached/detached to a visual tree and when its VisualChildren changes as follows:

```cs
using System.Diagnostics;
using Avalonia.Controls;

namespace TestAvalonia.Views;

public partial class TestUserControl : UserControl
{
    public TestUserControl()
    {
        InitializeComponent();

        VisualChildren.CollectionChanged += (sender, args) =>
        {
            Debug.WriteLine($"VisualChildren.CollectionChanged: {args.Action}");
        };

        AttachedToVisualTree += (sender, args) =>
        {
            Debug.WriteLine($"AttachedToVisualTree");
        };

        DetachedFromVisualTree += (sender, args) =>
        {
            Debug.WriteLine($"DetachedFromVisualTree");
        };
    }
}
```

During reproduction, the following will be observed:
```sh
// click - Step 2 of repro steps
AttachedToVisualTree
VisualChildren.CollectionChanged: Add

// click - Step 3 of repro steps
DetachedFromVisualTree

// click - Step 4 of repro steps
AttachedToVisualTree
VisualChildren.CollectionChanged: Remove

// click
DetachedFromVisualTree

// click
AttachedToVisualTree

// click
DetachedFromVisualTree

// click
AttachedToVisualTree
```
> I added annotation of when a click happens and formatted it a bit

As can be seen here, the visual children is somehow cleared when the user control is supposed to show again, and the children is never added back.

The repro project also contains some commented code in `MainWindow.axaml` and `MainWindow.axaml.cs` that changes the behaviour of the issue.

----

*MainWindow.axaml*
```xaml
    <Grid RowDefinitions="*,Auto">
        <!-- <TransitioningContentControl Name="PART_Panel" Grid.Row="0"/> -->
        <ContentControl Name="PART_Panel" Grid.Row="0"/>
        <Button Content="Switch view" Click="HandleButtonClick" Grid.Row="1"/>
    </Grid>
```
Using a `ContentControl` instead of `TransitioningContentControl` makes it so the visual children is never cleared, thus "fixing" the issue, but at the expense of transitions.

----

*MainWindow.axaml.cs*
```cs
    private Control[] _controls = new Control[]
    {
        new TextBlock()
        {
            Text = "Hello World! 1"
        },
        new TestUserControl(),
        new ContentControl()
        {
            Content = new TextBlock()
            {
                Text = "Hello World! 2"
            }
        }
    };
```

Making the `TransitioningContentControl` encounter another content control makes this weird behavior of it showing and not showing.

```
// click - Hello from user control!
AttachedToVisualTree
VisualChildren.CollectionChanged: Add

// click - Hello World! 2
DetachedFromVisualTree

// click - Hello World! 1

// click - Hello from user control! - invisible
AttachedToVisualTree
VisualChildren.CollectionChanged: Remove

// click - Hello World! 2
DetachedFromVisualTree

// click - Hello World! 1

// click - Hello from user control!
AttachedToVisualTree
VisualChildren.CollectionChanged: Add

... loops
```

### PR DESCRIPTION
**Pull request by danielmayost**
## What does the pull request do?
There are two problems with `TransitioningContentControl`:
- Duplicated logical children, this is because `ContentPresenter` manages his host's `LogicalChildren` (in this case TCC), and in addition `ContentControl` also manages `LogicalChildren`, which causes logical children to be duplicated.
- The first `ContentPresenter` not register the TCC as its Host, which causes it to not manage the `LogicalChildren` of the TCC.

We need to decide who will manage the logical children, CC (ContentControl) or TCC, in this case it is clear that TCC should do it, because CC is only adapted to one child, whereas TCC can have two (during animation).

~~The implementation was carried out by adding a `BypassLogicalChildrenManagement` property to the CC that allows the controls inheriting from it to manage the `LogicalChildren` themselves, in this case, TCC does not need to do anything, his presenters do it for him.~~ see comment below.

The second problem is a bug in the `RegisterContentPresenter` implementation that returned `false` even if Presenter is `PART_ContentPresenter`, which causes the Host registration to be skipped:
https://github.com/AvaloniaUI/Avalonia/blob/124f124617ffcd07da78e980ed04c0e6df595102/src/Avalonia.Controls/Presenters/ContentPresenter.cs#L702-L706

The fixes is pretty obvious.

## Checklist

- [x] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

## Breaking changes
No.

## Fixed issues
Fixes #12158

### CODE REVIEW

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0037671-beta`. (feed url: https://nuget-feed-all.avaloniaui.net/v3/index.json) [PRBUILDID]

File: src/Avalonia.Controls/ContentControl.cs [hunk 1]
```diff
@@ -95,6 +95,17 @@ public VerticalAlignment VerticalContentAlignment
...
+        /// <summary>
+        /// Determine whether <see cref="ContentControl"/> manage his LogicalChildren 
+        ///(default behavior) himself, or leaves the management to the inherited control.
+        /// </summary>
+        /// <remarks>
+		/// The default value is false, So the <see cref="ContentControl"/> manages itself, 
+        /// if you want to bypass this behavior and manage LogicalChildren yourself, set 
+        /// the <see cref="BypassLogicalChildrenManangment"/> to true.
+		/// </remarks>
+        protected virtual bool BypassLogicalChildrenManangment => false;
```
Comment (MrJul) [inline src/Avalonia.Controls/ContentControl.cs:107]: Typo: should be `Management`

**Comment by MrJul** [comment]
Nice one!

`ContentPresenter` automatically setting the `ContentControl` logical children makes sense, but it's really easy to forget about this relationship.

I think it would be nice to have debug warnings in the future when a logical child is added to a parent while it already has one. Even if there are valid use cases for this (as discussed with grokys previously in #10413), it could help catching this kind of problem sooner.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0037696-beta`. (feed url: https://nuget-feed-all.avaloniaui.net/v3/index.json) [PRBUILDID]

**Comment by X9VoiD** [comment]
Though I don't particularly agree with how it is fixed, can confirm it works! 👍 

The `ContentChanged` could probably be a `protected override`.

**Comment by danielmayost** [comment]
> The `ContentChanged` could probably be a `protected virtual`.

Then when we override it in the inheriting class without implementing anything, it won't be so clear why. I thought that way it would be clearer.

**Comment by MrJul** [comment]
> Then when we override it in the inheriting class without implementing anything, it won't be so clear why. I thought that way it would be clearer.

A simple comment could work. Alternatively, `ContentControl` can be changed to also use `OnPropertyChanged`, with `TransitioningContentControl` doing the same without calling the base for `ContentProperty` (with a comment, still). No new public members this way.

**Comment by danielmayost** [comment]
> A simple comment could work.

IMO it doesn't look aesthetic, but you decide.

> Alternatively, `ContentControl` can be changed to also use `OnPropertyChanged`, with TransitioningContentControl doing the same without calling the base for ContentProperty (with a comment, still).

Oh, this is a bad idea IMO, for two reasons:
- Avalonia's analyzer doesn't like it (actually throws a compilation error), to overriding `OnPropertyChanged` and not calling the base class.
- This is a problematic practice, you can't know who is in the hierarchy chain that `OnPropertyChanged` must be called, if you don't want to call the one in front of you it doesn't mean you don't want to call the one that's twice behind it.

**Comment by MrJul** [comment]
I think there's a misunderstanding here, I was definitely not advocating on *never* calling the `base.OnPropertyChanged`, that would indeed break everything! Sorry if I wasn't clear.

Not calling the base was only for `ContentProperty` in TCC, and after checking the code it's already the very case today:

```
protected override void OnPropertyChanged(AvaloniaPropertyChangedEventArgs change)
{
	if (change.Property == ContentProperty)
	{
		UpdateContent(true);
		return;
	}

	base.OnPropertyChanged(change);
}
```

(`ContentControl` would need to use `OnPropertyChanged` instead of `AddClassHandler`.)

**Comment by danielmayost** [comment]
Got it, so what's better?
I think the second option is better (OnPropertyChanged).

**Comment by MrJul** [comment]
Yes, that's a general problems with overrides and I agree it can be a pain when you're class 3 and need to repeat a behavior from class 1 that have been inhibited by class 2. Unfortunately C# doesn't allow calling an arbitrary ancestor's virtual method non-virtually (it's possible in IL).

In this particular case, the `Content` property is defined by `ContentControl`: `TransitioningContentControl` being a direct inheritor, there's no class before or between, plus both classes are in the same codebase: everything is in control. (The problem might apply to a potential inheritor of `TransitioningContentControl`, but I don't think that's a scenario we need to support.)

To me it's a win as it avoids defining a specific public flag that will only be used by `TransitioningContentControl` in practice. That said, I'll let an OG team member decides what's best.

**Comment by danielmayost** [comment]
@MrJul No matter, It's done, can it be merged now?

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0038197-beta`. (feed url: https://nuget-feed-all.avaloniaui.net/v3/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0038428-beta`. (feed url: https://nuget-feed-all.avaloniaui.net/v3/index.json) [PRBUILDID]

**Review by MrJul — APPROVED** [review]
LGTM, thanks!

**Pull request merged**